controllers/topology/internal/mergepatch/mergepatch.go

```diff
@@ -124,25 +135,25 @@ type applyPathOptionsOutput struct {
 }
 
 // applyPathOptions applies all the options acting on path level; currently it removes from the patch diffs not
-// in the allowed paths, filters out path to be ignored and enforce authoritative paths.
+// in the allowed paths, filters out path to be ignored and enforce authoritative and managed paths.
 // It also returns a flag indicating if the resulting patch has spec changes or not.
 func applyPathOptions(in *applyPathOptionsInput) (*applyPathOptionsOutput, error) {
 	twoWayPatchMap := make(map[string]interface{})
 	if err := json.Unmarshal(in.twoWayPatch, &twoWayPatchMap); err != nil {
 		return nil, errors.Wrap(err, "failed to unmarshal two way merge patch")
 	}
 
-	// Enforce changes from the authoritative patch when required.
+	// Enforce changes from authoritative patch when required (authoritative and managed paths).
 	// This will override instance specific fields for a subset of fields,
 	// e.g. machine template metadata changes should be reflected into generated objects without
 	// accounting for instance specific changes like we do for other maps into spec.
-	if len(in.options.authoritativePaths) > 0 {
+	if len(in.options.authoritativePaths) > 0 || len(in.options.managedPaths) > 0 {
 		authoritativePatchMap := make(map[string]interface{})
 		if err := json.Unmarshal(in.authoritativePatch, &authoritativePatchMap); err != nil {
 			return nil, errors.Wrap(err, "failed to unmarshal authoritative merge patch")
 		}
 
-		for _, path := range in.options.authoritativePaths {
+		for _, path := range append(in.options.managedPaths, in.options.authoritativePaths...) {
```

Comment:
Following scenario:

Initial KCPTemplate contains:
```yaml
spec:
  template:
    spec:
      kubeadmConfigSpec:
        clusterConfiguration:
          controllerManager:
            extraArgs: { enable-hostpath-provisioner: 'true' }
```

KCP reconciled managed-field annotation contains: `kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.enable-hostpath-provisioner`

KCPTemplate is now updated to drop the entire controllerManager sub-object.

Data in mergePatch helper:
* managedPaths contains `kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.enable-hostpath-provisioner`
* authoritativePatchMap contains a (nested) key `kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs` with value `nil`

As we are only opinionated about `enable-hostpath-provisioner` and not the parent object, empty'ing the extraArgs object is not part of the patch

Reply:
tl;dr I think we're not able to delete objects when we previously had opinions about nested fields.

I think I don't get the issue with DockerClusterTemplate with the loadBalancer object because in that case it's actually an empty object in the DockerClusterTemplate

Reply:
Nice catch
I have implemented a fix for managing changes on high-level objects. TL;DR;
- if a parent object is dropped from a template, we are deleting selectively what was set in a previous reconciliation (preserving nested user-defined values if any)
- if a parent object is explicitly set to null, the desired state aligns (dropping nested user-defined values if any)